kep/9-capacity-scheduling-for-ml-dl-workloads-based-on-scheduler-framework/README.md

```diff
@@ -0,0 +1,208 @@
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+We assume two elastic quotas are defined: QuotaA (guaranteed:4, hard:6) and QuotaB (guaranteed:6, hard:8), the quota unit is the number of GPUs. UserA in green color consumes QuotaA and UserB in red color consumes QuotaB. The entire cluster has 10 GPUs available hence the sum of quota guaranteed is equal to the cluster capacity.
+
+Initially, UserA consumes 2 GPUs and UserB consumes 3 GPUs. Both consumptions are within the reserved "guaranteed" quota range.
+<div style="text-align: center">
+<img src="./1.png"/>
+</div>
+
+
+Next, UserA consumes 4 GPUs reaching the "guaranteed" quota of QuotaA. UserB still consumes 3 GPUs. The cluster still has sufficient GPU resources hence no resource stealing happens.
+<div style="text-align: center">
+<img src="./2.png"/>
+</div>
+
+
+Assuming UserA attempts to use the rest of 3 unused GPUs in the cluster,  he can only use 2 more GPUs since his quota "hard" is 6 GPUs. Note that these 2 GPUs are "borrowed" from UserB's "Guaranteed" quota since UserB does not use up it. 
+<div style="text-align: center">
+<img src="./3.png"/>
+</div>
+
+
+Later, if UserB demands 8 GPUs, 5 more compared to his current consumption, his request cannot be immediately satisfied since only one unused GPU left in the cluster. Since UserB's "guaranteed" quota is 6 GPUs, the preemption algorithm ensures that three GPUs will be returned from UserA.  However, no more GPUs can be assigned to UserB since UserA's "guaranteed" quota has to be honored as well. In the end, UserA consumes 4 GPUs and UserB consumes 6 GPUs, satisfying each user's "guaranteed" quota.
+<div style="text-align: center">
+<img src="./4.png"/>
+</div>
+
+## Design Details
+
+### Extention point
+#### Preempt
+(This work will be started after the preemption extension point of framework finished)
+
+In the current scheduling workflow, if a Pod fails in `Filter` phase because of `FitError`, it will enter the `Preempt` phase. Assuming the Preempt plugin is supported, we will make the following changes to support the “Queue” based preemption algorithm.
+
+The failed to schedule Pod is referred to as the “preemptor”. The original `selectVictimsOnNode` method selects all the pods with the lower priority than the preemptor’s priority as potential victims in a node. In this proposal, the victim selection logic needs to be changed in the following cases.
+If the preemptor belongs to a namespace without Queue defined, the preemption victims will be selected from other namespaces without Queue defined.
+If the preemptor belongs to a namespace with Queue defined:
+When preemptor.Request + Queue.allocated > Queue.min. The preemption victims will be selected from the pods belonging to the same namespace based on the Pod priority.
```

Comment:
Is queue.allocated from the API spec (e.g., status) or cache?

Reply:
> The preemption victims will be selected from the pods belonging to the same namespace based on the Pod priority.

The failure reason can be arbitrary, not particular with a scarce resource. So the algorithm described here (and below) seems to assume the failure reason is just because pods among the queues are competing, but what if it's not the case? In other words, (1) why preempting pods from the queues can resolve the unschedulable state, and (2) why preempting pods from the queues is optimal?

Reply:
> Is queue.allocated from the API spec (e.g., status) or cache?
-----
from cache. Thanks I add this in this kep.

Reply:
>  why preempting pods from the queues can resolve the unschedulable state
----
We have two different types of workloads: offline and online. The application of online workload doesn‘t support preemption, so they also doesn‘t support the `borrow` and `return` in ElasticQuota. So the min and max is equal in online workload. And the rest is offline workloads. Most of the failures of pods from offline workloads come from a scarce resource. So the preemption from the queues can resolve the unschedulable state.

> why preempting pods from the queues is optimal?
----
The pods from the queues maybe not optimal. But the optional range needs to be limited by Quota. 
If quota.allocated + pod.request > quota.min, the preemption will happen Intra-Quota.
if quota.allocated + pod.request < quota.min, the preemption will happen Inter-Quota.

@Huang-Wei